src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs

```diff
@@ -14,871 +13,607 @@
 namespace System.Net.Security
 {
     //
-    // The class maintains the state of the authentication process and the security context.
-    // It encapsulates security context and does the real work in authentication and
+    // The class does the real work in authentication and
     // user data encryption with NEGO SSPI package.
     //
     // This is part of the NegotiateStream PAL.
     //
-    internal class NegoState
+    internal partial class NegoState
     {
-        private const int ERROR_TRUST_FAILURE = 1790;   // Used to serialize protectionLevel or impersonationLevel mismatch error to the remote side.
-
-        private static readonly byte[] s_emptyMessage = new byte[0];
-        private static readonly AsyncCallback s_readCallback = new AsyncCallback(ReadCallback);
-        private static readonly AsyncCallback s_writeCallback = new AsyncCallback(WriteCallback);
-
-        private Stream _innerStream;
-        private bool _leaveStreamOpen;
+        internal IIdentity GetIdentity()
+        {
+            CheckThrow(true);
 
-        private Exception _exception;
+            IIdentity result = null;
+            string name = _context.IsServer ? _context.AssociatedName : _context.Spn;
+            string protocol = "NTLM";
 
-        private StreamFramer _framer;
-        private NTAuthentication _context;
+            protocol = _context.ProtocolName;
 
-        private int _nestedAuth;
+            if (_context.IsServer)
+            {
+                SecurityContextTokenHandle token = null;
+                try
+                {
+                    SecurityStatusPal status;
+                    SafeDeleteContext securityContext = _context.GetContext(out status);
+                    if (status != SecurityStatusPal.OK)
+                    {
+                        throw new Win32Exception((int)SslStreamPal.GetInteropFromSecurityStatusPal(status));
+                    }
 
-        internal const int MaxReadFrameSize = 64 * 1024;
-        internal const int MaxWriteDataSize = 63 * 1024; // 1k for the framing and trailer that is always less as per SSPI.
+                    // This will return a client token when conducted authentication on server side.
+                    // This token can be used for impersonation. We use it to create a WindowsIdentity and hand it out to the server app.
+                    Interop.SecurityStatus winStatus = (Interop.SecurityStatus)SSPIWrapper.QuerySecurityContextToken(
+                        GlobalSSPI.SSPIAuth,
+                        securityContext,
+                        out token);
+                    if (winStatus != Interop.SecurityStatus.OK)
+                    {
+                        throw new Win32Exception((int)winStatus);
```

Comment:
Please ignore my comment about Win32Exception (I see that this part won't change due to those changes). So @stephentoub , reason for not using CreateException.. is to avoid the mapping between Interop.SecurityStatus and SecurityStatusPal.